File: clipsync-win/ClipSync.Core/Clipboard/SuppressionPolicy.cs
using System;
using ClipSync.Settings;

namespace ClipSync.Clipboard;

/// What the suppression decision needs to know about the foreground: which
/// app was in front at a given instant. `ForegroundRing` and
/// `ForegroundTracker` both satisfy it, so tests can drive the decision from
/// a ring populated synthetically instead of from real windows.
public interface IForegroundSource
{
    AppIdentity? AppAt(DateTime utc);
}

/// The one decision the excluded-apps feature exists to make: given the
/// instant a copy happened, may the item be transmitted to peers?
///
/// Lives here, apart from the clipboard plumbing, because it is the
/// assertion the feature is judged on and it must be directly testable.
public static class SuppressionPolicy
{
    /// True when the item must NOT be transmitted.
    ///
    /// Fails open in every uncertain case: an unresolved source app, or any
    /// exception on the way to the answer, yields false (transmit). Silent
    /// non-delivery is a worse failure than a rare miss, and a confidently
    /// wrong identity is worse than none.
    ///
    /// `source` is the resolved app, or null when it could not be
    /// determined. It is presentation/logging data only — never clipboard
    /// content — and is set whether or not the item is suppressed.
    public static bool ShouldSuppress(IForegroundSource foreground, AppSettings settings,
                                      DateTime copiedAtUtc, out AppIdentity? source)
    {
        source = null;
        try
        {
            source = foreground.AppAt(copiedAtUtc);
            if (source is null) return false;
            return settings.IsExcluded(source);
        }
        catch (Exception ex)
        {
            Security.Log.Write($"SuppressionPolicy: decision failed, transmitting: {ex.GetType().Name}");
            return false;
        }
    }
}

File: clipsync-mac/Sources/ClipSync/Clipboard/SuppressionPolicy.swift
import Foundation

/// The one decision the excluded-apps feature exists to make: given the
/// polling window a copy fell in, may the item be transmitted to peers?
///
/// Lives apart from the clipboard plumbing because it is the assertion
/// the feature is judged on and it must be directly testable — and
/// because on Windows the same decision, buried inline, once failed
/// *closed* through a bare catch without anyone noticing. The
/// unresolvable case here is an explicit branch, not an exception path.
enum SuppressionPolicy {

    struct Decision {
        /// True when the item must NOT be transmitted.
        let suppress: Bool
        /// The excluded app that triggered suppression, or nil when
        /// transmitting. Presentation/logging data only — never content.
        let source: AppIdentity?
    }

    /// The macOS rule (deliberately stricter than Windows, where the
    /// exact copy timestamp is known): if **any** app that held focus
    /// during the polling window `(windowStart, windowEnd]` is excluded,
    /// suppress. The alternative — whoever is frontmost at tick time —
    /// leaks on a fast copy-then-switch, which is exactly the sequence of
    /// copying a password and switching to the app that needs it.
    ///
    /// Fails open in every uncertain case: an empty ring, an unresolved
    /// app, or a window predating everything retained all yield
    /// "transmit". Silent non-delivery is a worse failure than a rare
    /// miss.
    static func decide(ring: ForegroundRing, settings: AppSettings,
                       windowStart: Date, windowEnd: Date) -> Decision {
        let candidates = ring.appsIn(start: windowStart, end: windowEnd)
        for app in candidates where settings.isExcluded(app) {
            return Decision(suppress: true, source: app)
        }
        return Decision(suppress: false, source: nil)
    }
}
